src/app/clusters/operational-state-server/operational-state-server.cpp

```diff
@@ -60,139 +64,158 @@ CHIP_ERROR OperationalStateServer::Init()
+        mOperationalError.Set(to_underlying(ErrorStateEnum::kNoError));
+        ConcreteAttributePath path = ConcreteAttributePath(mEndpointId, mClusterId, Attributes::OperationalError::Id);
+        MatterReportingAttributeChangeCallback(path);
     }
-    else if (opState != to_underlying(OperationalStateEnum::kPaused))
+
+    uint8_t oldState  = mOperationalState;
+    mOperationalState = aOpState;
+    if (mOperationalState != oldState)
     {
-        delegate->HandlePauseStateCallback(err);
+        ConcreteAttributePath path = ConcreteAttributePath(mEndpointId, mClusterId, Attributes::OperationalState::Id);
+        MatterReportingAttributeChangeCallback(path);
     }
-    response.commandResponseState = err;
+    return CHIP_NO_ERROR;
+}
 
-    ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+DataModel::Nullable<uint8_t> Instance::GetCurrentPhase() const
+{
+    return mCurrentPhase;
 }
 
-void OperationalStateServer::HandleResumeState(HandlerContext & ctx, const Commands::Resume::DecodableType & req)
+uint8_t Instance::GetCurrentOperationalState() const
 {
-    ChipLogDetail(Zcl, "OperationalState: HandleResumeState");
-    Commands::OperationalCommandResponse::Type response;
-    Delegate * delegate = OperationalState::GetOperationalStateDelegate(mEndpointId, mClusterId);
-    GenericOperationalError err(to_underlying(ErrorStateEnum::kNoError));
+    return mOperationalState;
+}
 
-    VerifyOrReturn(delegate != nullptr, ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure));
-    uint8_t opState = delegate->GetCurrentOperationalState();
+void Instance::GetCurrentOperationalError(GenericOperationalError & error) const
+{
+    error.Set(mOperationalError.errorStateID, mOperationalError.errorStateLabel, mOperationalError.errorStateDetails);
+}
 
-    if (opState != to_underlying(OperationalStateEnum::kPaused) && opState != to_underlying(OperationalStateEnum::kRunning))
+void Instance::OnOperationalErrorDetected(const Structs::ErrorStateStruct::Type & aError)
+{
+    ChipLogDetail(Zcl, "OperationalStateServer: OnOperationalErrorDetected");
+    // Set the OperationalState attribute to Error
+    if (mOperationalState != to_underlying(OperationalStateEnum::kError))
     {
-        err.Set(to_underlying(ErrorStateEnum::kCommandInvalidInState));
+        mOperationalState          = to_underlying(OperationalStateEnum::kError);
+        ConcreteAttributePath path = ConcreteAttributePath(mEndpointId, mClusterId, Attributes::OperationalState::Id);
+        MatterReportingAttributeChangeCallback(path);
     }
-    else if (opState == to_underlying(OperationalStateEnum::kPaused))
+
+    // Set the OperationalError attribute
+    if (aError.errorStateID != mOperationalError.errorStateID ||
+        !aError.errorStateLabel.Value().data_equal(mOperationalError.errorStateLabel.Value()) ||
+        !aError.errorStateDetails.Value().data_equal(mOperationalError.errorStateDetails.Value()))
```

Comment:
If the label or details of either one are not present, this will crash, no?

Reply:
Fixed [here](https://github.com/project-chip/connectedhomeip/pull/28577/commits/3dfe2c2e25435d70254427c679995edb55c1bef2).